external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/CachedKafkaConsumer.scala

```diff
@@ -310,62 +308,45 @@ private[kafka010] object CachedKafkaConsumer extends Logging {
 
   private lazy val cache = {
     val conf = SparkEnv.get.conf
-    val capacity = conf.getInt("spark.sql.kafkaConsumerCache.capacity", 64)
-    new ju.LinkedHashMap[CacheKey, CachedKafkaConsumer](capacity, 0.75f, true) {
-      override def removeEldestEntry(
-        entry: ju.Map.Entry[CacheKey, CachedKafkaConsumer]): Boolean = {
-        if (entry.getValue.inuse == false && this.size > capacity) {
-          logWarning(s"KafkaConsumer cache hitting max capacity of $capacity, " +
-            s"removing consumer for ${entry.getKey}")
-          try {
-            entry.getValue.close()
-          } catch {
-            case e: SparkException =>
-              logError(s"Error closing earliest Kafka consumer for ${entry.getKey}", e)
-          }
-          true
-        } else {
-          false
+    val capacityConfigString: String = "spark.sql.kafkaConsumerCache.capacity"
+    val capacity = conf.getInt(capacityConfigString, 64)
+    val duration = conf.getTimeAsMs("spark.sql.kafkaConsumerCache.timeout", "10m")
+    val removalListener = new RemovalListener[CacheKey, CachedKafkaConsumer] {
+      override def onRemoval(n: RemovalNotification[CacheKey, CachedKafkaConsumer]): Unit = {
+        n.getCause match {
+          case RemovalCause.SIZE =>
+            logWarning(
```

Comment:
It shouldn't really be a normal operation.  If capacity is smaller than the number of partitions that are regularly being assigned to a given node, it's going to kill performance due to recreating consumers every batch.